activemq-broker/src/main/java/org/apache/activemq/broker/region/Topic.java

```diff
@@ -826,14 +830,51 @@ protected void dispatch(final ConnectionContext context, Message message) throws
     }
 
     private final AtomicBoolean expiryTaskInProgress = new AtomicBoolean(false);
-    private final Runnable expireMessagesWork = new Runnable() {
-        @Override
-        public void run() {
-            List<Message> browsedMessages = new InsertionCountList<Message>();
-            doBrowse(browsedMessages, getMaxExpirePageSize());
+    private final Runnable expireMessagesWork = () -> {
+        try {
+            final TopicMessageStore store = Topic.this.topicStore;
+            if (store != null && store.getType() == StoreType.KAHADB) {
```

Comment:
Would it be suitable to check the class instanceof/isassingable/isinstance vs having the type flag? Anyone extending KahaDB would still have this functionality.

Reply:
@mattrpav - I actually forgot that even ignoring the proxy stuff (Because you could check delegates) it's not possible to do an instance of or class check. The code lives in the activemq-broker module and that module has no dependency on the data stores like KahaDB. So the KahaDB specific interfaces are not even on the classpath at compilation time.

Reply:
That makes sense, thanks.